pkg/genericcli/context.go

```diff
@@ -0,0 +1,682 @@
+	"github.com/metal-stack/metal-lib/pkg/multisort"
+	"github.com/metal-stack/metal-lib/pkg/pointer"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	KeyName           = "name"
+	KeyAPIURL         = "api-url"
+	KeyAPIToken       = "api-token"
+	KeyDefaultProject = "default-project"
+	KeyTimeout        = "timeout"
+	KeyActivate       = "activate"
+	KeyProvider       = "provider"
+	KeyConfig         = "config"
+
+	defaultConfigName = "config.yaml"
+)
+
+var (
+	// errorNotImplemented for all functions that are not implemented yet
+	errorNotImplemented = fmt.Errorf("not implemented yet")
+)
+
+// Contexts contains all configuration Contexts
+type Contexts struct {
+	CurrentContext  string     `json:"current-context" yaml:"current-context"`
+	PreviousContext string     `json:"previous-context" yaml:"previous-context"`
+	Contexts        []*Context `json:"contexts" yaml:"contexts"`
+}
+
+type Context struct {
+	Name           string         `json:"name" yaml:"name"`
+	APIURL         *string        `json:"api-url,omitempty" yaml:"api-url,omitempty"`
+	APIToken       string         `json:"api-token" yaml:"api-token"`
+	DefaultProject string         `json:"default-project" yaml:"default-project"`
+	Timeout        *time.Duration `json:"timeout,omitempty" yaml:"timeout,omitempty"`
+	Provider       string         `json:"provider" yaml:"provider"`
+}
+
+type ContextConfig struct {
+	BinaryName    string
+	ConfigName    string
+	ConfigDirName string
+	Fs            afero.Fs
+
+	// I/O
+	DescribePrinter       func() printers.Printer
+	ListPrinter           func() printers.Printer
+	In                    io.Reader
+	Out                   io.Writer
+	ProjectListCompletion cobra.CompletionFunc
+}
+
+type cliWrapper struct {
+	cfg *ContextConfig
+}
+
+type contextUpdateRequest struct {
+	Name string
+}
+
+// NewContextCmd creates the context command tree using genericcli
+func NewContextCmd(c *ContextConfig) *cobra.Command {
+	// TODO check nils
+	c.ConfigName = cmp.Or(c.ConfigName, string(defaultConfigName))
+	c.Out = cmp.Or(c.Out, io.Writer(os.Stdout))
+	c.In = cmp.Or(c.In, io.Reader(os.Stdin))
+	c.Fs = cmp.Or(c.Fs, afero.NewOsFs())
+
+	if c.ConfigDirName == "" {
+		panic(fmt.Errorf("no config directory name provided"))
+	}
+
+	wrapper := &cliWrapper{
+		cfg: c,
+	}
+
+	cmd := NewCmds(&CmdsConfig[
+		*Context,
+		*contextUpdateRequest,
+		*Context,
+	]{
+		GenericCLI:      NewGenericCLI(wrapper),
+		BinaryName:      c.BinaryName,
+		Singular:        "context",
+		Plural:          "contexts",
+		Description:     "Manage CLI contexts. A context defines the connection properties (API URL, token, etc.) for a backend. Use \"-\" to switch to the previously used context.",
+		Aliases:         []string{"ctx"},
+		Args:            []string{KeyName},
+		Sorter:          contextSorter(),
+		DescribePrinter: c.DescribePrinter,
+		ListPrinter:     func() printers.Printer { return newPrinterFromCLI(c) },
```

Comment:
```suggestion
		ListPrinter:     c.ListPrinter,
```

We already have one. No need to create a new one.